### PR TITLE
Support css-modules

### LINKED ISSUE #4306: Storybook default webpack config does not support CRA v2 w/ CSS modules
### Bug or support request summary

From https://storybook.js.org/configurations/custom-webpack-config/:

> The default Webpack config of Storybook is balanced for a medium-size project (specially created with Create React App) or a library.

I am assuming that this means that Storybook's default webpack config should be able to handle a new project created by CRA. CRA v2 now has [opt-in support for CSS modules](https://reactjs.org/blog/2018/10/01/create-react-app-v2.html#whats-new). When I created a new CRA project using CRA v2.0.4, I used CSS modules to style a component; however, the component styles do not appear when displaying the component in Storybook.

If this means that I need to extend Storybook's webpack config to add CSS module support, please close this issue. However, it's unclear if Storybook is meant to work out-of-the-box with CRA if this is the case. I can make a PR to add documentation around this if Storybook will not add CSS module support out of the box like CRA v2 has.

### Steps to reproduce

I have a reproducible example in [this repo](https://github.com/matthamil/storybook-css-module-CRA2).

1. Create a project with `npx create-react-app my-app`
1. Add Storybook to the project with `getstorybook`
1. Create a component that imports a CSS module.
1. Render the component in a Storybook story.

If you import the component that uses CSS modules into `App.js` and run `yarn run start`, you can see that the styles are applied correctly. However, if you run Storybook with `yarn run storybook`, the styles do not apply.

### Please specify which version of Storybook and optionally any affected addons that you're running

package.json
```json
{
  "name": "storybook-css-module",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "react": "^16.5.2",
    "react-dom": "^16.5.2",
    "react-scripts": "2.0.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "storybook": "start-storybook -p 9009 -s public",
    "build-storybook": "build-storybook -s public"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": [
    ">0.2%",
    "not dead",
    "not ie <= 11",
    "not op_mini all"
  ],
  "devDependencies": {
    "@storybook/react": "^3.4.11",
    "@storybook/addon-actions": "^3.4.11",
    "@storybook/addon-links": "^3.4.11",
    "@storybook/addons": "^3.4.11",
    "babel-core": "^6.26.3",
    "babel-runtime": "^6.26.0"
  }
}

```

### Affected platforms 

Tested in Chrome v69

### Screenshots / Screencast / Code Snippets (Optional)

Repo: https://github.com/matthamil/storybook-css-module-CRA2

Text component when running `yarn run start`:

image

Text component when running `yarn run storybook`:

image

### Work summary

Add CSS module support to Storybook's webpack config.

### Where to start

Take a look at `./src/Text.js` and `./src/Text.module.css`. The CSS module includes one class `.text` which applies a red color and 2rem font-size to whatever element is using `.text`.

### Acceptance criteria

Default storybook webpack config can handle CSS modules out of the box.

### Who to contact

anyone familiar with Storybook's default webpack config.

### PR DESCRIPTION
**Pull request by chadfawcett**
Issue: Resolve #4306

## What I did

Created a react webpack preset that pulls style rules from `create-react-app`'s congis.
* Support for css modules (`*.module.css`)
* Support for sass (`scss` and `sass`)
* Support for sass modules (`*.module.scss` and `*.module.sass`)

~> The default Webpack config of Storybook is balanced for a medium-size project (specially created with Create React App) or a library.~

~I know @ndelangen mentioned that this statement is a little outdated, but I feel this change should be okay as it shouldn't cause any harm adding css modules as is.~

~@igor-dv I chose to propose the simpler option first, as `presets` aren't well defined yet, and I saw @ndelangen was working on separating the manager and preview webpack processes. Either of these may result in a better implementation of the CRA webpack config in the future.~

~I can work on implementing the `presets` approach if this method is not sufficient. What are your thoughts?~

## How to test

You can update `examples/cra-kitchen-sink` to the following in order to test (example code taken from initial reporting issue):

```css
//  App.module.css (New file)
.text {
  color: red;
  font-size: 2rem;
}
```

```js
//  App.js (Existing file)
...
import styles from './App.module.css'

const App = () => (
  ...
    <h2 className={styles.text}>Welcome to React</h2>
  ...
)
```

Then run `yarn storybook` from the `examples/cra-kitchen-sink` repo and the "Welcome to React" will be red.

screen shot 2018-10-12 at 9 31 25 pm

_Does this need a new example in the kitchen sink apps?_
Maybe? I can add something similar to the code above if we want it committed into the repo.

### CODE REVIEW

**Comment by chadfawcett** [comment]
@gabrielcsapo if it was isolated at the example level, wouldn’t that mean it was an end user override similar to the workaround proposed in the original issue?

**Comment by ndelangen** [comment]
@chadfawcett is correct here.

I think the best approach would be to narrow this into a preset specific for either React of CRA.

I'm sure @igor-dv will be able to help with this.

**Comment by chadfawcett** [comment]
I’ve got a fairly good idea of how to implement the preset, but it’s a lot more work to merge the default and cra configs. So I thought I’d propose this solution first 😜

**Comment by ndelangen** [comment]
I fear this will not work well with other frameworks.

**Review by pksunkara — DISMISSED** [review]

**Comment by chadfawcett** [comment]
@ndelangen Would it cause anymore issues with other frameworks than the existing `css` rules?

**Review by igor-dv — CHANGES_REQUESTED** [review]
I think we need to be very careful with the changes like this since you are adding it to the **default** webpack config and it may really break other usages/frameworks without the real benefit to them. Also, the convention of `file.module.css` - I am not sure how much this is common (for example I don't use it =). So, who is really winning from this change?

> I’ve got a fairly good idea of how to implement the preset, but it’s a lot more work to merge the default and cra configs

Integration with CRA is acutaly the thing we need to do 🤷‍♂️ . Also, I assume we do not need to merge the full config, but only styling, at least for now IMO (I didn't see any other issues that are not related to styling). 

P.S sorry for being unresponsive, I am returning from the vacation tomorrow =)

**Comment by chadfawcett** [comment]
@igor-dv Okay so it seems the consensus is that updating the default config directly is bad, and we should use a preset. There are two approaches we could take:

1. Detect cra and automatically merge in parts of the cra webpack config ([I could use your feedback on my proposed method of detection](https://github.com/storybooks/storybook/issues/4306#issuecomment-429182221))
2. Apply css module support to **all** `@storybook/react`. This would just be a css modules preset that we add to `storybook/app/react/src/server/options.js`

For reference, I figured #4298 could also be fixed if we merged more than just the style rules.

Thoughts?

**Comment by igor-dv** [comment]
I vote for **1** =) Anything neede to be merged should be merged

**Comment by ndelangen** [comment]
Would be cool if we'd load the actual webpack config CRA uses, and merge the relevant bits

**Comment by gabrielcsapo** [comment]
I agree 1 is the way to go

**Comment by Hypnosphi** [comment]
> I fear this will not work well with other frameworks.

Why? CSS Modules are completely framework agnostic

**Comment by chadfawcett** [comment]
@Hypnosphi @ndelangen That's what I thought. It may not be the best implementation, but it's and easy win that shouldn't cause issue with anything else.

Should I update the regex to support sass/scss and then call this PR done? Leave the CRA preset to a later PR?

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
@igor-dv do you agree to merge this as is?

**Comment by igor-dv** [comment]
Guys, if you find this useful, I will not be a blocker =)

**Comment by ndelangen** [comment]
@kylemh You're working on postcss stuff, which I think is related.

WDYT?

**Comment by kylemh** [comment]
Postcss loading would come in after CSS loader so this doesn't funk with compatibility here.

I'm interested in knowing how this regex works though... Does CRA map it's CSS files to `.module.css`?

If we're trying to make changes to this to get CRA to work out of the box, and CRA is indeed appending `.modules` to style files, I bet we could easily provide good defaults for Sass loader as well.

File: lib/core/src/server/config/webpack.config.default.js
Comment (kylemh) [inline lib/core/src/server/config/webpack.config.default.js]: we'll still want our default postcss in action here.

You could save line 23-34 as `defaultPostcssLoader` variable and then spread it into both style loading blobs.

**Review by kylemh — CHANGES_REQUESTED** [review]

**Comment by chadfawcett** [comment]
@kylemh [css modules](https://github.com/css-modules/css-modules) is optional in CRA. The `.modules.css` extension is manually set by the end developer. 

This is CRA's webpack config https://github.com/facebook/create-react-app/blob/f89385f7438f39750b762590ebd9729534e30190/packages/react-scripts/config/webpack.config.dev.js#L306-L321

File: lib/core/src/server/config/webpack.config.default.js
Comment (chadfawcett) [reply lib/core/src/server/config/webpack.config.default.js]: Good catch! CRA has a [`getStyleLoaders`](https://github.com/facebook/create-react-app/blob/f89385f7438f39750b762590ebd9729534e30190/packages/react-scripts/config/webpack.config.dev.js#L45) function and I missed the `postcss-loader` portion of it.

Comment (igor-dv) [reply lib/core/src/server/config/webpack.config.default.js]: That is why we need to integrate with cra properly, and not mimicking it.

Comment (chadfawcett) [reply lib/core/src/server/config/webpack.config.default.js]: Agreed, integrating is the better approach. This is just following how it's currently implemented.

**Comment by kylemh** [comment]
Very interesting. Since it's a forced paradigm, I don't mind copying it wholesale. I think this would be better served as an aspect of the Storybook React builder, rather than webpack defaults, but I also don't mind doing this in the interim, given that it's boxed out via a required file suffix.

**Comment by chadfawcett** [comment]
I wasn't trying to keep the scope small. @ndelangen mentioned above it could be nice to expand the regex to capture `scss` and `sass`, which I agree with but was waiting to hear on whether we wanted to merge this in at all. I could also resolve the svg import issues as mentioned a few times but again, I wanted to keep the scope relatively small. If we went with this simple approach I was going to submit a separate PR for the svg. If we are wanting to not merge this and go with the presets, then I'd look at tackling everything at once.

**Comment by kylemh** [comment]
The SVG issue should definitely be separate. I think you can keep the PR's scope small and accomplish all the necessary work to have style config parity with CRA.

File: lib/core/src/server/config/webpack.config.default.js
Comment (kylemh) [inline lib/core/src/server/config/webpack.config.default.js]: I think you meant to keep this

Comment (chadfawcett) [reply lib/core/src/server/config/webpack.config.default.js]: No, I updated it to use `postcss-preset-env`.

https://github.com/storybooks/storybook/pull/4405/files/4a6fd1baf91c7f19324bc735c26aa981f4ac8410#diff-ba9b40ce49bd5df0209abd09fc032f58R28

Comment (kylemh) [reply lib/core/src/server/config/webpack.config.default.js]: Very cool - haven't heart of this!

All of the below if for my incoming PR:

So if this is based upon `.browserlistrc`... What happens when their coverage disagrees with our hard-coded autoprefixer config?

Lastly, I couldn't find anything on supporting PostCSS within CRA 2.0

I'm certain the answer should be: "leave it to them to figure out in full control mode", but should we do anything further for the rare situation where CRA users eject and come up with their own PostCSS config? It should be simple enough, and it further justifies the need to modularize our increasingly complex style-related webpack config logic.

Comment (chadfawcett) [reply lib/core/src/server/config/webpack.config.default.js]: I was just trying to stick to what the CRA config had. `postcss-preset-env` will look at the `browserlist` set in the `package.json`, so not sure where this differs from what was there before.

I agree the styles-related webpack config needs to be modularized. To be honest I picked up this PR as I thought it was going to be simple and would allow me to get my toes wet with the codebase. It's turning out to be a bit bigger of a change haha. The more I add the more I'm not sure if I'm conflicting with some of the other apps.

Wouldn't #4411 support PostCSS within CRA 2.0?

File: lib/core/src/server/config/webpack.config.default.js
Comment (chadfawcett) [reply lib/core/src/server/config/webpack.config.default.js]: Sure!

File: lib/core/src/server/config/webpack.config.default.js
Comment (chadfawcett) [reply lib/core/src/server/config/webpack.config.default.js]: Are you suggesting I move both the helper function (`getStyleLoaders`) and the style rules below into a new file? Or just the helper function?

File: lib/core/src/server/config/webpack.config.default.js
Comment (chadfawcett) [reply lib/core/src/server/config/webpack.config.default.js]: Thanks, nitpicking is fine, I'm just trying to follow along lol.

It looks like the post-css is clashing with the `marko` package. Both mine and yours are failing that build. I'm not familiar with the `marko` framework so I'm not 100% sure but it looks like it's because we removed the `postcss: {}` line from the default config. https://github.com/storybooks/storybook/blob/1f127f0b1d0cdd71708ed4864bc79055b1796f4f/lib/core/src/server/config/webpack.config.default.js#L24

I added that back in and now the `marko` package seems to work.

Comment (kylemh) [reply lib/core/src/server/config/webpack.config.default.js]: Nice usage of the new suggestion feature 😉 

I haven't dived in at all, but I'm certain we can provide an escape hatch for one consumer if we're providing a green path for another.

This could turn into a slippery slope, but we are trying to support one of the biggest consumer bases via this PR.

@ndelangen  - know any Marko experts?

File: lib/core/src/server/config/styleWebpackRules.js
Comment (igor-dv) [inline lib/core/src/server/config/styleWebpackRules.js]: Does it mean we need to have a `sass-loader` (and `node-sass` probably) in `@storybook/core` ?

Comment (ndelangen) [reply lib/core/src/server/config/styleWebpackRules.js]: node sass is a unmanaged dependency for CRA2 vai sass-loader:
https://github.com/webpack-contrib/sass-loader

So we do not need a dependency on `node-sass`

File: lib/core/src/server/config/styleWebpackRules.js
Comment (chadfawcett) [inline lib/core/src/server/config/styleWebpackRules.js]: Do I need to update the `angular-cli_config.js` to now also filter out `/\.(scss|sass)$/`?

https://github.com/storybooks/storybook/blob/80ec5726b68aea0dff1245642a888e43c07daaf3/app/angular/src/server/angular-cli_config.js#L114

Comment (igor-dv) [reply lib/core/src/server/config/styleWebpackRules.js]: It should be filtered, yeah. It was supposed to be done in #4343.

File: lib/core/src/server/config/styleWebpackRules.js
Comment (igor-dv) [inline lib/core/src/server/config/styleWebpackRules.js]: Wasn't this PR about to include by default only `.module.**` ? because it will probably conflict now with https://github.com/storybooks/storybook/blob/master/app/angular/src/server/framework-preset-angular.js#L35

Comment (chadfawcett) [reply lib/core/src/server/config/styleWebpackRules.js]: I updated the angular style rules filter as suggested in #4343 so it shouldn't conflict.

Though the suggested solution of `rule.test.test()` doesn't work well with [all valid options](https://webpack.js.org/configuration/module/#condition) of `test`. Working on resolving that now.

Comment (igor-dv) [reply lib/core/src/server/config/styleWebpackRules.js]: The thing is when people are not using angular-cli (it's rare though), they will not have this filtering rules logic.

Comment (igor-dv) [reply lib/core/src/server/config/styleWebpackRules.js]: We can have all these CRA rules , if moving from changing the `core` defaults to changing the `app/react` defaults

Comment (chadfawcett) [reply lib/core/src/server/config/styleWebpackRules.js]: I know 😔. I was hoping to find an interim solution. Maybe changing the `app/react` defaults won’t be as hard now that I’ve had a lot more exposure to the codebase haha.

File: lib/core/src/server/config/styleWebpackRules.js
Comment (kylemh) [reply lib/core/src/server/config/styleWebpackRules.js]: Maybe it's worth adding some error logging here.

Something to the extent of: If you've seen this error, we're assuming that you're using CRA 2.0 and have `node-sass` already installed as a dependency.

Comment (chadfawcett) [reply lib/core/src/server/config/styleWebpackRules.js]: @kylemh Where would this error message be implemented?

File: lib/core/src/server/config/styleWebpackRules.js
Comment (kylemh) [reply lib/core/src/server/config/styleWebpackRules.js]: I don't think we should provide Sass support beyond `.modules.s(c|a)ss`

Even though you resolved the Angular issue, I think we're going beyond the original scope of the issue in trying to support _anybody_'s Sass usage.

File: lib/core/src/server/config/styleWebpackRules.js
Comment (kylemh) [reply lib/core/src/server/config/styleWebpackRules.js]: I just spent like 30 minutes checking into custom error handling on a per-loader basis, but it looks like there isn't anything of the sort. 🤷‍♂️

**Comment by chadfawcett** [comment]
@igor-dv and @kylemh Can I please get a review of the new preset approach?

File: app/react/src/server/framework-preset-cra-styles.js
Comment (igor-dv) [inline app/react/src/server/framework-preset-cra-styles.js]: `webpackFinal` kicks in after all the webpack configuration manipulations. I will commit my stashed changes into your PR to fix the webpackFinal phase. It will solve https://github.com/storybooks/storybook/issues/4296#issuecomment-431095849 and the similar issues for CRA that will possibly come.

File: app/react/src/server/cra_config.js
Comment (igor-dv) [inline app/react/src/server/cra_config.js]: Why do you need to save `normalizeCondition` as a module state? Can't we just import it?

File: app/react/src/server/cra_config.js
Comment (igor-dv) [inline app/react/src/server/cra_config.js]: Same here. Do we need it as a dep?

File: app/react/src/server/cra_config.js
Comment (igor-dv) [inline app/react/src/server/cra_config.js]: please extract to `getCraWebpackConfig`

File: app/react/src/server/cra_config.js
Comment (igor-dv) [inline app/react/src/server/cra_config.js]: Do we really need this, though?

File: lib/core/src/server/core-preset-webpack-custom.js
Comment (igor-dv) [inline lib/core/src/server/core-preset-webpack-custom.js]: `webpackFinal` will be called twice. One for the `defaultConfig` and second for the `baseConfig`. That is how it worked before the `presets` feature 🤷‍♂️

File: app/react/src/server/framework-preset-cra-styles.js
Comment (chadfawcett) [reply app/react/src/server/framework-preset-cra-styles.js]: When I used `webpack()` none of the default styles had been loaded in yet, so based off of the angular-cli presets I chose `webpackFinal()`. My understanding was that the `presets` aren't documented yet so I wasn't sure if there were other options.

Is `webpackFinal()` the right choice here now that you've pushed your fix?

File: app/react/src/server/cra_config.js
Comment (chadfawcett) [reply app/react/src/server/cra_config.js]: We need `normalizeCondition` in both `getStyleRules` and `applyCRAWebpackConfig` but I only wanted to import it once. My rationale was that if we import at the beginning of the file, and `react-scripts` **isn't** a dependency, then we could get errors.

Though now that I think about it, this is more necessary for the `mini-css-extract-plugin` below. Webpack **shouldn't** ever error since it's a dependency of `@storybook/core`. However, if we import it directly, I think `webpack` should be added as a dependency of `@storybook/react`.

What are you thoughts?

File: app/react/src/server/cra_config.js
Comment (chadfawcett) [reply app/react/src/server/cra_config.js]: We don't need this as a dependency since we're first checking if `react-scripts` is installed. If so, then `mini-css-extract-plugin` will be installed.

File: app/react/src/server/cra_config.js
Comment (chadfawcett) [reply app/react/src/server/cra_config.js]: Nope! I just thought minification would be good for production.

Also, if we don't want to include this plugin, then we have to filter out the related rule from all of the style rules. Adding the plugin was less work than removing it.
https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/config/webpack.config.prod.js#L65-L71

File: app/react/src/server/framework-preset-cra-styles.js
Comment (igor-dv) [reply app/react/src/server/framework-preset-cra-styles.js]: Yes, that's the place for now, and since it is still internal , we can change this whenever we need =)

**Comment by kylemh** [comment]
@chadfawcett I'll have a review in by EOD PST

File: app/react/src/server/cra_config.js
Comment (igor-dv) [reply app/react/src/server/cra_config.js]: I don't see problems with adding it as a dep

**Review by kylemh — APPROVED** [review]
😍

**Comment by kylemh** [comment]
This preset approach is very interesting.

I wonder if we could set ground rules and documentation about adding new ones.

Most importantly, I wonder if we could give the user control in using presets thus removing the need for us to handle edge-cases. Presets can only be selected under X assumptions - otherwise, you're on your own.

File: app/react/package.json
Comment (igor-dv) [reply app/react/package.json]: webpack is a direct dep in a bunch of the SB packages (including `core`). It will be complicated to make it a peer (like for supporting different major version of webpack) 

CC #4371, #4430

File: app/react/src/server/cra_config.js
Comment (igor-dv) [reply app/react/src/server/cra_config.js]: So you can require it already in the `applyCRAWebpackConfig` func and remove the state from the module (IMO Node caches imported modules, so you don't need to care about this issue).

**Review by igor-dv — APPROVED** [review]
Now I am happy 👍

**Comment by igor-dv** [comment]
Almost forgot. Do we have any CLI tests testing this setup?

File: app/react/src/server/cra_config.js
Comment (chadfawcett) [reply app/react/src/server/cra_config.js]: @kylemh This preset checks for `react-scripts` greater than v2. So it would not do anything for CRAv1.

**Comment by shilman** [comment]
@chadfawcett @igor-dv @ndelangen would love to get this into an RC -- any chance we can merge this today?

**Comment by igor-dv** [comment]
The CLI tests could be added after

**Comment by chadfawcett** [comment]
Sorry I was in the middle of responding to comments when GitHub started having issues yesterday. Glad my latest push synced in time.

@igor-dv As for the CLI tests, all of the fixtures have React 15 as dependencies. This brought to light that I needed to test that `react-scripts` was at least v2, but that meant the CLI tests were no longer testing this new preset.

Should the fixtures be updated to React 16? Or should additional fixtures be created?

**Comment by igor-dv** [comment]
I think React 15 is not supported now, I am not sure how it works. But yeah, you can just create a new one.

File: app/react/src/server/cra_config.js
Comment (igor-dv) [inline app/react/src/server/cra_config.js]: On a second thought, I think this can make problems if 'mini-css-extract-plugin' is not installed as the root package in node_modules 🤔

Comment (igor-dv) [reply app/react/src/server/cra_config.js]: Maybe we need to do something like this instead ?

```js
const miniCssExtractPlugin = craWebpackConfig.plugins.find(plugin => plugin.constructor.name === 'MiniCssExtractPlugin');
```

Comment (chadfawcett) [reply app/react/src/server/cra_config.js]: I think you were right initially. Since we check to make sure `react-scripts` is installed before calling `applyCRAWebpackConfig` (I Just noticed this has all caps for CRA), then we know `mini-css-extract-plugin` is present.

Comment (Hypnosphi) [reply app/react/src/server/cra_config.js]: If we want to use it, we have to depend on it

Comment (Hypnosphi) [reply app/react/src/server/cra_config.js]: > we know mini-css-extract-plugin is present.

there's no guarantee that it's available in root `node_modules` not in `node_modules/react-scripts/node_modules`

Comment (igor-dv) [reply app/react/src/server/cra_config.js]: Looks like #4524 is broken now because of this

Comment (chadfawcett) [reply app/react/src/server/cra_config.js]: #4534

File: app/react/package.json
Comment (Hypnosphi) [inline app/react/package.json]: It should be `"4.0.0-rc.1"`

**Comment by tirli** [comment]
Seems like scss is still not working under windows out of the box. We tested it on Linux and it works OK, but with the same config(just initialized) on windows it fails with error `You may need an appropriate loader to handle this file type`.

Tested in the combination of storybook 4.0.12 and CRA 2.1.1

**Pull request merged**
